File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
from typing import List
from collections import Counter

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        n = len(s)
        m = len(words)
        wl = len(words[0])

        permutations = Counter(words)

        result = []
        for i in range(n - m * wl + 1):
            seen = Counter()
            j = 0
            while j < m:
                word = s[i + j * wl:i + (j + 1) * wl]
                if word in permutations:
                    seen[word] += 1
                    if seen[word] > permutations[word]:
                        break
                else:
                    break
                j += 1
            if j == m:
                result.append(i)

        return result

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        int n = s.length();
        int m = words.size();
        int wl = words[0].length();

        unordered_map<string, int> permutations;
        for (int i = 0; i < m; i++) {
            permutations[words[i]]++;
        }

        vector<int> result;
        for (int i = 0; i <= n - m * wl; i++) {
            unordered_map<string, int> seen;
            int j = 0;
            for (; j < m; j++) {
                string word = s.substr(i + j * wl, wl);
                if (permutations.find(word) != permutations.end()) {
                    seen[word]++;
                    if (seen[word] > permutations[word]) break;
                } else break;
            }
            if (j == m) result.push_back(i);
        }
        return result;
    }
};
